### ISSUE TITLE
shaded javac should set Automatic-Module-Name and shade service loader entries

### ISSUE THREAD

**Issue by sormuras** [google/google-java-format#266]
When integrating `google-java-format` as a library (into an application that lives on the `module-path`) there are some issues. Especially with the 3rd-party dependencies `google-java-format` relies on.

One issue, namely the dependency to a certain `error-prone/javac` release, could be tackled by using `javac` provided by the current runtime system.

Copied from https://github.com/forax/pro/pull/34#issuecomment-373315521
_For the javac-9-dev-r3297-1-shaded.jar, it would be cool if google-java-format had an option to use javac through the service interface as we do in the compiler plugin instead of using its own version._

**Comment by cushon** [comment]
Can you expand on what the issues are? The motivation for the shaded javac is that GJF uses internal javac APIs that change between releases. Supporting arbitrary javac versions doesn't seem that useful, because it will still only work with that latest version. What problems are you seeing with `-all-deps.jar` in that PR?

**Comment by sormuras** [comment]
I'm just in the process of rebasing the PR to gjf 1.5 and current pro master. There are two ways to include gjf into an external application, like pro. a) Use the 3 single jars or b) use the `-all-deps.jar`.

a)
Quote: _I didn't managed to make modules via _module-fixer_ out of neither the "-all-deps" distro nor the 3 single jars:  `google-java-format-1.3.jar`, `guava-19.0.jar` and `javac-9-dev-r3297-1-shaded.jar`_

Not all problems I faced with the _none_ `-all-deps.jar` way come to my mind now.
Forax' pro tool runs entirely on the module-path. iirc, gjf poses no problem. Especially after providing an `Automatic-Module-Name`. Guava was not behaving well back then, and javac-... had some split-package issues.

b)
My solution then was to fallback to the command line and wrap/control the external process within a pro plugin. This worked then and does work now -- but it is not really ... integrated.

So, the idea is to implement the `formatter` plugin for pro using approach a). Pro always comes with a current/latest javac version bundled in its jlink-ed runtime image. It would be nice if gfj could use that javac via the tool interface first. Maybe gjf could introduce an option to use the default "javac" tool?

**Comment by sormuras** [comment]
Did my explanation clarify the use case?

**Comment by cushon** [comment]
Having an option to use other javacs is complicated by the way we're doing the shading. It's possible, but I'm not really understanding why it's necessary. It sounds like (a) works with `-all-deps` now?

**Comment by sormuras** [comment]
(a) works and was merged by @forax lately. Resorting to spawn another VM doesn't feel like it is really integrated, though.

**Comment by cushon** [comment]
Sorry, I'm still not clear on what the underlying issue is here, or why it was necessary to run GJF in a separate process.

Do you have a self-contained repro that shows the issue you saw using GJF as a library from a modularized application?

**Comment by sormuras** [comment]
> Do you have a self-contained repro that shows the issue you saw using GJF as a library from a modularized application?

Should be available a local branch... if not, I'll prepare one.

**Comment by sormuras** [comment]
### DIY

I compiled a minimal demo. Checkout ["bach"](https://github.com/sormuras/bach) and run `Build03GoogleJavaFormatAsLibrary` by invoking `jshell build-03-google-java-format-as-library.jsh` in the `demo/03-google-java-format-as-library` folder. JDK-10 is required.

### TL;DR

`javac-shaded-9+181-r4173-1.jar` does not work well as an automatic module. It'd be cool, if `google-java-format` had only deps to `javax.tools...` instead of `org.openjdk...`. One could add `requires java.compiler;` to the application's module descriptor.

### Setup

- `src/application/foo/Foo.java`  is the application stub using gjf as a libray.
```
package foo;

import com.google.googlejavaformat.java.Formatter;

public class Foo {

  public static void main(String... args) throws Exception {
    var sourceString = "package foo; enum Bar {A,B,C}";
    var formattedSource = new Formatter().formatSource(sourceString);
    System.out.println(formattedSource);
  }
}
```

- `src/application/module-info.java`  reads `com.google.googlejavaformat`.
```
module application {
  requires com.google.googlejavaformat;
}
```

### Module- and Class-path

Working on the `--module-path` like a charm:
- `google-java-format-1.6-20180410.041304-23.jar`
- `guava-24.1-jre.jar`

Needs to live on the `--class-path`:
- `javac-shaded-9+181-r4173-1.jar`
...because:
-  JPMS can't convert "file name" -> "module name"
*  doesn't provide Automatic-Module-Name in it's MANIFEST.MF
-  rename manually to "errorprone.jar" and put on the module-path yields:
```Error occurred during initialization of boot layer
java.lang.module.FindException: Unable to derive module descriptor for
              .bach\module-path\errorprone.jar
          Caused by: java.lang.module.InvalidModuleDescriptorException: Provider class
              com.sun.tools.javac.platform.JDKPlatformProvider not in module
```

**Comment by cushon** [comment]
Thanks! I understand the problem better now.

I was able to reproduce the `InvalidModuleDescriptorException` by creating a GJF jar with an `Automatic-Module-Name`. The issue there is a `META-INF/services` didn't get shaded:

```
$ unzip -p .bach/class-path/javac-shaded-9+181-r4173-1.jar META-INF/services/com.sun.tools.javac.platform.PlatformProvider
com.sun.tools.javac.platform.JDKPlatformProvider
```

Fixing that to refer to the shaded classes allows your repro to work with GJF on the module path.

I'll include an `Automatic-Module-Name` and fix the service loading issue in the next `error-prone-javac` release.

**Comment by cushon** [comment]
The `Automatic-Module-Name` part happened in https://github.com/google/google-java-format/commit/4d728e7c, and we're no longer using a shaded javac so the service loader stuff should be fixed also: https://github.com/google/google-java-format/commit/3c191c1326b77aabbf2a48c3edc19b628f4327d6.
